File: ds/fenwickTree-1.py
class FenwickTree:
    '''
    Fenwick Tree which supports Point update & Range queries
    '''
    def __init__(self, arr):
        n = len(arr)
        self.ft = [0]*(n)
        # update with all the values
        for i in range(1, len(self.ft)):
            self.update(i, arr[i])

    @staticmethod
    def LSOne(s):
        return s & (-s)

    def update(self, i, v):
        while i < len(self.ft):
            self.ft[i] += v
            i += FenwickTree.LSOne(i)

    def __query(self, r):
        sm = 0
        while r:
            sm += self.ft[r]
            r -= FenwickTree.LSOne(r)
        return sm

    def queryRange(self, l, r):
        # special handling if l is 0
        return self.__query(r) - (0 if l == 1 else self.__query(l-1))

if __name__ == "__main__":
    arr = [0] + [5, 9, 6, 1, 8, 4, 3, 7, 2, 8]
    print(sum(arr))
    ft = FenwickTree(arr)
    print(ft.queryRange(1, 10))
    print(ft.queryRange(4,6))
    ft.update(5, 2)
    print(ft.queryRange(4,6))

File: java/masterchief/src/main/java/ds/fenwickTree/FenwickTree1.java
package ds.fenwickTree;

/**
 * This is a Fenwick tree implementation
 * with Point Updates & Range Queries
 */
public class FenwickTree1 {
    private int[] ft;

    public FenwickTree1(int [] arr, int size) {
        this.ft = new int[size + 1];
        // initialize ft with array elements
        for (int i = 1; i < this.ft.length; i++) {
            update(i, arr[i-1]);
        }
    }

    /**
     * update the fenwick tree
     * @param p index position in the array in question; 1-based indexing
     * @param v value in the array at that position
     */
    public void update(int p, int v) {
        for (; p < this.ft.length; p += p & -p) {
            this.ft[p] += v;
        }
    }

    /**
     * get the sum of the range inclusive
     * 1-based indexing
     * @param left
     * @param right
     */
    public int query(int left, int right) {
        return query(right) - query(left);
    }

    private int query(int p) {
        int sum = 0;
        for (; p > 0; p -= p & -p) {
            sum += this.ft[p];
        }
        return sum;
    }
}
